tests/base/src/test/java/org/keycloak/tests/admin/authz/fgap/RoleResourceTypeEvaluationTest.java

```diff
@@ -0,0 +1,175 @@
+
+        UserPolicyRepresentation onlyMyAdminUserPolicy = createUserPolicy(realm, client, "Only My Admin User Policy", myadmin.getId());
+        // we need to be able to list client scopes
+        createPermission(client, myclient.getId(), AdminPermissionsSchema.CLIENTS.getType(), Set.of(VIEW), onlyMyAdminUserPolicy);
+
+        // create a client-scope
+        ClientScopeRepresentation clientScope = new ClientScopeRepresentation();
+        clientScope.setName("my-client-scope");
+        clientScope.setProtocol("openid-connect");
+        try (Response response = realm.admin().clientScopes().create(clientScope)) {
+            assertThat(response.getStatus(), equalTo(Response.Status.CREATED.getStatusCode()));
+            clientScope.setId(ApiUtil.handleCreatedResponse(response));
+            realm.cleanup().add(r -> r.clientScopes().get(clientScope.getId()).remove());
+        }
+
+        // we don't have permissions to map roles to a client scope so the list of available roles should be empty
+        ClientScopeResource clientScopeResource = realmAdminClient.realm(realm.getName()).clientScopes().get(clientScope.getId());
+        List<RoleRepresentation> availableRoles = clientScopeResource.getScopeMappings().realmLevel().listAvailable();
+        assertThat(availableRoles, empty());
+
+        // grant the permission to map all roles to client scopes
+        createAllPermission(client, rolesType, onlyMyAdminUserPolicy, Set.of(MAP_ROLE_CLIENT_SCOPE));
+
+        availableRoles = clientScopeResource.getScopeMappings().realmLevel().listAvailable();
+        assertThat(availableRoles, not(empty()));
+    }
+
+    @Test
+    public void testMapCompositeRoleAllRoles() {
+        UserRepresentation myadmin = realm.admin().users().search("myadmin").get(0);
+
+        // create a role and sub-role
+        RoleRepresentation role = new RoleRepresentation();
+        role.setName("myRole");
+        realm.admin().roles().create(role);
+        realm.cleanup().add(r -> r.roles().get("myRole").remove());
+
+        RoleRepresentation subRole = new RoleRepresentation();
+        subRole.setName("mySubRole");
+        realm.admin().roles().create(subRole);
+        subRole = realm.admin().roles().get("mySubRole").toRepresentation();
+        realm.cleanup().add(r -> r.roles().get("mySubRole").remove());
+
+        // the following operation should fail as the permission wasn't granted yet
+        try {
+            realmAdminClient.realm(realm.getName()).roles().get("myRole").addComposites(List.of(subRole));
+            fail("Expected exception wasn't thrown.");
+        } catch (Exception ex) {
+            assertThat(ex, instanceOf(ForbiddenException.class));
+        }
+
+        // TODO for some reason, addComposites endpoint is guarded by "manage-realm" role which is then again checked in canMapComposite(), making MAP_ROLE_COMPOSITE permission redundant
```

Comment:
By redundant you mean when the admin user is also granted with `manage-realm` role?

Reply:
Yeah, which has to be granted to an admin otherwise `addComposites` endpoint throws 403 immediately.